usr.bin/mdo/mdo.c

```diff
@@ -19,52 +20,277 @@
 static void
 usage(void)
 {
-	fprintf(stderr, "usage: mdo [-u username] [-i] [--] [command [args]]\n");
-	exit(EXIT_FAILURE);
+	fprintf(stderr,
+		"Usage: mdo [options] [--] [command [args...]]\n"
+		"\n"
+		"Options:\n"
+		"  -u <user>       Target user (name or UID)\n"
+		"  -i              Only change UID, skip groups\n"
+		"  -g <group>      Override primary group (name or GID)\n"
+		"  -G <g1,g2,...>  Set supplementary groups (name or GID list)\n"
+		"  -s <mods>       Modify supplementary groups using:\n"
+		"                   +group to add, -group to remove, @ to reset\n"
+		"\n"
+		"Advanced UID/GID overrides:\n"
+		"  -U <ruid>       Set real UID\n"
+		"  -R <svuid>      Set saved UID\n"
+		"  -E <euid>       Set effective UID\n"
+		"  -P <rgid>       Set real GID\n"
+		"  -Q <svgid>      Set saved GID\n"
+		"\n"
+		"  --print-rule	  Print the actual rules of transition in mac.do.rules format\n"
+		"  -h              Show this help message\n"
+		"\n"
+		"Examples:\n"
+		"  mdo -u alice id\n"
+		"  mdo -u 1001 -g wheel -G staff,operator /bin/sh\n"
+		"  mdo -u bob -s @,+wheel,+operator /usr/bin/id\n"
+		"  mdo -E 1002 -R 1003 -U 1004 /bin/id\n"
+	);
+	exit(1);
 }
 
 int
 main(int argc, char **argv)
 {
-	struct passwd *pw;
+	struct passwd *pw = NULL;
 	const char *username = "root";
+	const char *primary_group = NULL;
+	const char *supp_groups_str = NULL;
+	const char *group_mod_str = NULL;
 	struct setcred wcred = SETCRED_INITIALIZER;
 	u_int setcred_flags = 0;
 	bool uidonly = false;
 	int ch;
 
-	while ((ch = getopt(argc, argv, "u:i")) != -1) {
+	gid_t gid = -1;
+	bool override_gid = false;
+
+	bool supp_reset = false;
+	gid_t *supp_add = NULL, *supp_rem = NULL;
+	size_t add_count = 0, rem_count = 0;
+
+	const char *ruid_str = NULL, *svuid_str = NULL, *euid_str = NULL, *rgid_str = NULL, *svgid_str = NULL;
```

Comment:
This line is probably too long.  Have one line for user IDs, and another one for group IDs.